### PR TITLE
Try to fix flake TestServerReconcilePod

### PR DESCRIPTION
**Pull request by sword-jin**
**Please provide a description of this PR:**

This PR tries to fix https://github.com/istio/istio/issues/48247

From the error log here https://storage.googleapis.com/istio-prow/logs/unit-tests_istio_postsubmit/1733170613905788928/build-log.txt,

```
    informers_test.go:207: found diff:   bool(
        - 	true,
        + 	false,
          )
        
        Left: true
        Right: false
add app-2
add app-2
del app-2
```

this failure could be reproduced when executing thousands of times `go test -race -run ^TestServerReconcilePod$ istio.io/istio/cni/pkg/ambient -v --count=1000 > debug.log`.

Just realized the previous test code has bug, it should return false in the for loop,

```go
assert.EventuallyEqual(t, func() bool {
	skippedEvents := 0
	for i, event := range c.expectedEvents {
           ...
           +++ return false +++
	}
	return true
}, true, retry.Timeout(3*time.Second))
```

This missed `return false` hided a few problems.

Anyway, the times and order of received events could not both be guaranteed, I removed assert all events and we could focus on which event must be received.

Also, the `DeletionTime` is always be nil when using fake client.

### CODE REVIEW

**Comment by hanxiaop** [comment]
@sword-jin I suggest that you apply to become an Istio member; then you won't need the `ok-to-test` anymore.

**Comment by sword-jin** [comment]
> @sword-jin I suggest that you apply to become an Istio member; then you won't need the `ok-to-test` anymore.

That's great, my team allows me about 1 hour which is dedicated to istio.

**Comment by hanxiaop** [comment]
> > @sword-jin I suggest that you apply to become an Istio member; then you won't need the `ok-to-test` anymore.
> 
> That's great, my team allows me about 1 hour which is dedicated to istio.

See https://github.com/istio/community/pull/1274 as an example.

File: cni/pkg/ambient/informers_test.go [hunk 1]
```diff
@@ -184,7 +182,6 @@ func TestServerReconcilePod(t *testing.T) {
 
 		t.Run(c.name, func(t *testing.T) {
 			ctx, cancel := context.WithCancel(ctx)
-			defer cancel()
```
Comment (hzxuzhonghu) [inline cni/pkg/ambient/informers_test.go:187]: if the move is not a must, please restore

Comment (sword-jin) [reply cni/pkg/ambient/informers_test.go:187]: I call it manually in line 199 to make sure all the events handled by informer.

File: cni/pkg/ambient/informers_test.go [hunk 2]
```diff
@@ -267,7 +264,7 @@ func (h *fakePodReconcileHandler) delPodFromMesh(pod *v1.Pod, event controllers.
 
 	h.events = append(h.events, podEvent{
 		event: "del",
-		pod:   pod,
+		pod:   pod.DeepCopy(),
```
Comment (hzxuzhonghu) [inline cni/pkg/ambient/informers_test.go:267]: Instead of deepcopy here, i would suggest finding where it is mutated and prefer to fix it there.

Comment (sword-jin) [reply cni/pkg/ambient/informers_test.go:267]: Agree with you, any idea to reproduce this kind of error in local. I tested it under the high cpu stress(99.9%), and also passed.

File: cni/pkg/ambient/informers_test.go (hunk 1 above)
Comment (hzxuzhonghu) [reply cni/pkg/ambient/informers_test.go:187]: cancel often comes along with context init. what is the intention to stop kubeclient before

File: cni/pkg/ambient/informers_test.go (hunk 2 above)
Comment (hzxuzhonghu) [reply cni/pkg/ambient/informers_test.go:270]: go test -race -count

Comment (sword-jin) [reply cni/pkg/ambient/informers_test.go:270]: It reproduced when set count to a big number.

Just realized the previous code has bug, it should return false in the forloop,

```go
assert.EventuallyEqual(t, func() bool {
	skippedEvents := 0
	for i, event := range c.expectedEvents {
           ...
           +++ return false +++
	}
	return true
}, true, retry.Timeout(3*time.Second))
```

This missed `return false` hided a frw problems.

**Review by hzxuzhonghu — APPROVED** [review]

**Pull request merged**